src/java/main/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -990,6 +992,27 @@ private void sendPing() {
         private boolean saslLoginFailed = false;
 
         private void startConnect(InetSocketAddress addr) throws IOException {
+            boolean canonicalize = true;
+            try {
+                canonicalize = Boolean.parseBoolean(System.getProperty(ZK_SASL_CLIENT_CANONICALIZE_HOSTNAME, "true"));
+            } catch (IllegalArgumentException ea) {
+                //ignored ...
+            }
+
+            if (canonicalize) {
```

Comment:
You only want to do this if SASL client is enabled. Hence the name of the switch contains `sasl.client`.